packages/network/src/connection/NodeWebRtcConnection.ts

```diff
@@ -0,0 +1,318 @@
+        if (this.isFinished) {
+            // already closed, noop
+            return
+        }
+
+        this.isFinished = true
+
+        if (err) {
+            this.logger.warn('conn.close(): %s', err)
+        } else {
+            this.logger.trace('conn.close()')
+        }
+
+        if (this.connectionEmitter) {
+            this.connectionEmitter.removeAllListeners()
+        }
+
+        if (this.dataChannelEmitter) {
+            this.dataChannelEmitter.removeAllListeners()
+        }
+
+        if (this.connection) {
+            try {
+                this.connection.close()
+            } catch (e) {
+                this.logger.warn('conn.close() errored: %s', e)
+            }
+        }
+
+        if (this.dataChannel) {
+            try {
+                this.dataChannel.close()
+            } catch (e) {
+                this.logger.warn('dc.close() errored: %s', e)
+            }
+        }
+
+        this.dataChannel = null
+        this.connection = null
+
+        this.doClose()
+    }
+
+    getBufferedAmount(): number {
+        try {
+            return this.dataChannel!.bufferedAmount().valueOf()
+        } catch (err) {
+            return 0
+        }
+    }
+
+    getMaxMessageSize(): number {
+        try {
+            return this.dataChannel!.maxMessageSize().valueOf()
+        } catch (err) {
+            return 1024 * 1024
+        }
+    }
+ 
+    isOpen(): boolean {
+        try {
+            return this.dataChannel!.isOpen()
+        } catch (err) {
+            return false
+        }
+    }
+
+    private onStateChange(state: string): void {
+        this.logger.trace('conn.onStateChange: %s -> %s', this.lastState, state)
+
+        this.lastState = state
+
+        if (state === 'disconnected' || state === 'closed') {
+            this.close()
+        } else if (state === 'failed') {
+            this.close(new Error('connection failed'))
+        } else if (state === 'connecting') {
+            // reset timeout on connecting
+            clearTimeout(this.connectionTimeoutRef!)
+            this.connectionTimeoutRef = setTimeout(() => {
+                if (this.isFinished) { return }
+                this.logger.warn(`connection timed out after ${this.newConnectionTimeout}ms`)
+                this.close(new Error(`timed out after ${this.newConnectionTimeout}ms`))
+            }, this.newConnectionTimeout)
+        }
+    }
+
+    private onGatheringStateChange(state: string): void {
+        this.logger.trace('conn.onGatheringStateChange: %s -> %s', this.lastGatheringState, state)
+        this.lastGatheringState = state
+    }
+
+    private onDataChannel(dataChannel: DataChannel): void {
+        this.setupDataChannel(dataChannel)
+        this.logger.trace('connection.onDataChannel')
+        this.openDataChannel(dataChannel)
+    }
+
+    private onLocalDescription(description: string, type: DescriptionType): void {
+        this.emit('localDescription', type, description)
```

Comment:
Let's consider moving the event emission to base class `WebRtcConnection`.

I.e. replace this with
```ts
private onLocalDescription(description: string, type: DescriptionType): void {
        super.onLocalDescription(type, description) // super equivalent?
}
```

And then in the base class `WebRtcConnection`:

```ts
protected onLocalDescription(description: string, type: DescriptionType): void {
    this.emit('localDescription', type, description)
}
```

Additionally we need to add a comment to the class JSDoc of `WebRtcEndpoint` that says that "implementers of sub classes need to take care to call protected method onLocalDescription whenever their underlying webrtc library generates a local description".